_posts/2023-04-30-writing-code-with-llm-dont-forget-kani.md

```diff
@@ -0,0 +1,306 @@
+There were two failed checks.
+The first one `Failed Checks: attempt to add with overflow` says that the addition `x + 1` may overflow.
+The other one says that equivalent check may fail---the two implementations are actually not equivalent.
+
+
+To better understand the bug, we asked Kani to produce us a counterexample with the `--conrete-playback` command (discussed in [our previous blog](https://model-checking.github.io/kani-verifier-blog/2022/09/22/internship-projects-2022-concrete-playback.html◊))
+```
+> cargo kani --enable-unstable --concrete-playback=print | grep "//" | awk '{print $2}'
+2147483647
+```
+The counterexample is when x equals to `2147483647` (`i32::MAX`).
+
+
+The examples above illustrate that, although code generation is efficient and accessible, it's crucial to exercise caution and avoid blindly trusting the generated code, as it may contain errors.
+We believe there are two factors contribute to this unreliability.
+Firstly, the training corpus for most LLMs is unverified code.
+Secondly, text prompts can be ambiguous, leading to the production of flawed code.
+Fortunately, program verifiers such as Kani can provide valuable information about generated code, such as verification results, and counterexamples.
+Armed with this information, developers can locate and fix bugs more effectively.
+But what about the potential for language models like ChatGPT to leverage such insights to improve code generation accuracy?
+It's an intriguing question that merits further exploration.
+
+
+
+
+## Verified Code Generation with Verifiers in the Loop
+The last task we explore in this blog is: can we combine Kani and ChatGPT to generate correct and verified code?
+
+#### Example 1.c
+Continue following Example 1.b.
+Recall that we found the overflow error of the expression `a + b` in the generated code with Kani.
+We shared this information with ChatGPT.
+```
+Q: The expression (a + b) / 2 may overflow.
+```
+ChatGPT agreed with us and fix the overflow with a new implementation using `checked_add`:
+``` rust
+fn integer_average(num1: i32, num2: i32) -> i32 {
+    match num1.checked_add(num2) {
+        Some(sum) => sum / 2,
+        None => {
+            // Handle overflow gracefully
+            println!("Overflow occurred!");
+            std::i32::MAX // Fallback value
+        }
+    }
+}
+```
+
+We reran Kani on this new implementation.
+Kani verified that this new version is correct!
+However, the function become no longer complete that computes the average for all i32 integers as we expected.
+For example, it does not compute the average for the input `integer_average(std::i32::MAX, 1)`.
+So, we tried again with the following prompt to generate a complete version of `integer_average`.
+```
+Q: Can you give me a fix that does not throw any error?
+```
+This time, ChatGPT find us an implementation that can also be verified by Kani, and accept all i32 integers as input.
+```
+fn integer_average(num1: i32, num2: i32) -> i32 {
```

Comment:
Maybe close the loop by running Kani once again and show that the verification succeeds.